### PR TITLE
SAK-31533: General configuration table for content review activities

### PR DESCRIPTION
**Pull request by plukasew**
https://jira.sakaiproject.org/browse/SAK-31533

### CODE REVIEW

**Comment by plukasew** [comment]
This is my first pass at the general content review activity configuration table discussed on the mailing list. It consists of the table and a dao. The configuration data is not currently exposed through the ContentReviewService interface, but implementations can use the dao to store and access their configuration data.

I'm still unclear on the design goals for ContentReviewService so I don't know whether or not to expose this configuration data in the larger API. I'm not sure if the API is supposed to be a "lowest common denominator" spec, or if it should know about all possible options a review service could have and allow querying of capabilities.

If it is the former, it probably makes sense to expose the general config table directly, with getConfig/setConfig type methods in ContentReviewService. This would however expose implementation details and would allow bugs such as a tool writing the wrong config data.

If it is the latter, it might be better to keep the config an implementation detail and just expose methods like "supportsOperationA" and then "doOperationA" and "getConfigForOperationA" (if required).

The two concrete examples I have are for the new Turnitin LTI integration. The first is for the external assignment identifier on the Turnitin side. As it stands now, there is no reason for a tool like Assignments to know this id directly, so exposing it could potentially lead to bugs if someone were to change the value in Assignments code.

The other example is an identifier for the LTI tool instance that corresponds to the assignment. This is purely a Turnitin-LTI-implementation-specific property, yet the Assignments tool needs to know the value, either directly or indirectly, so it can provide a link to directly access the Turnitin assignment through the LTI tool . The current way is to add a method to the ContentReviewService API called getLTIAccess(), which is meant to return a URL that contains this LTI id. Clearly this is meaningless for other content review services and they shouldn't have to implement it. So we could expose the configuration directly and let the Assignments code read the URL from the config (assignments would have to then know the name of the config property), or we could generalize the API with methods like supportsDirectAccessToExternalActivity() and getDirectAccessUrlForExternalActivity().

File: content-review/impl/dao/src/main/java/org/sakaiproject/contentreview/dao/ContentReviewActivityConfigDao.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	 * Creates or updates an activity config entry for the given key (unique combination of name, toolId, activityId, and providerId)
+	 * @param name the name of the configuration property
+	 * @param value the value of the configuration property
+	 * @param activityId the unique identifier for the activity (ex: an assignment id)
+	 * @param toolId the unique identifier for the Sakai tool the activity belongs to (ex: sakai.assignment.grades)
+	 * @param providerId the unique identifier for the content review provider this configuration is for
+	 * @param overrideIfSet if an entry for the given key already exists, override the existing value
+	 * @return true if a new entry was created or an existing entry was modified
+	 */
+	public boolean createOrUpdateEntry(String name, String value, String activityId, String toolId, int providerId, boolean overrideIfSet)
```
Comment (ern) [inline content-review/impl/dao/src/main/java/org/sakaiproject/contentreview/dao/ContentReviewActivityConfigDao.java]: It probably makes sense if the logic of this method was in a service.

**Comment by ern** [comment]
Hi Paul,

You pretty much have the pattern.

In the API are the POJO's and related hbm's. These are considered an "Exposed Data Model".

In the IMPL are the DAO's and they should only be concerned with things like "get this item where id equals x", "get this item where category equals x and not category equals y". The only other thing needed is some Hibernate boiler plate, HibernateCommonDao which your DAO should extend.

If the service is public its interface will reside in the API otherwise it will reside in the IMPL along with its implementation.

The service implementation contains all the logic and typically should be representative of the domain that the service is representing. The service also contains all the transaction management, since one service method may call several DAO methods and wrap all of those in a single transaction.

I totally agree with you that the less a tool (in this case Assignments) knows about the internals the better, and it should never expand beyond the exposed POJO or a service method.

**Comment by ern** [comment]
As an aside and probably little to do with your PR here...
I would like to say that there probably needs some new generic support for LTI based content review integrations. Trying to take an api based content review integration and repurposing it for LTI probably doesn't make much sense, however abstracting the commonalities so that they may be shared is a good design choice.

**Comment by plukasew** [comment]
The implementation details probably need to be revised since I wrote this a while back and there were actually no services included, only DAOs, which is why I stuck all the logic there. I'll revisit this when we are ready to contribute our 11.x LTI integration work back to master.

**Comment by bjones86** [comment]
Decision on the core team call today was to close this for the time being. When we (Western) get around to contributing all our work for the TII LTI integration to 12, we can revisit this.

**Pull request closed without merging**